File: 01-Programming-Languages/Java/Data-Structures/Queue/QueueUsingStacks.java
/*
 * Problem: Implement a Queue using two stacks.
 * Approach: Use one stack for enqueue, another for dequeue (amortized O(1)).
 * Time Complexity: O(1) amortized per operation
 * Space Complexity: O(n)
 * Example: enqueue(1), enqueue(2), dequeue() -> 1, enqueue(3), dequeue() -> 2
 */

import java.util.*;

public class QueueUsingStacks {
    private Deque<Integer> in = new ArrayDeque<>();
    private Deque<Integer> out = new ArrayDeque<>();

    public void enqueue(int x) {
        in.push(x);
    }

    public int dequeue() {
        if (out.isEmpty()) {
            while (!in.isEmpty()) out.push(in.pop());
        }
        return out.pop();
    }

    public int peek() {
        if (out.isEmpty()) {
            while (!in.isEmpty()) out.push(in.pop());
        }
        return out.peek();
    }

    public boolean isEmpty() {
        return in.isEmpty() && out.isEmpty();
    }

    public static void main(String[] args) {
        QueueUsingStacks q = new QueueUsingStacks();
        q.enqueue(1);
        q.enqueue(2);
        System.out.println(q.dequeue());
        q.enqueue(3);
        System.out.println(q.dequeue());
        System.out.println(q.dequeue());
    }
}

File: 01-Programming-Languages/Kotlin/Data-Structures/Queue/queue_using_stacks.kt
/*
 * Problem: Implement a Queue using two stacks.
 * Approach: Use one stack for enqueue, another for dequeue (amortized O(1)).
 * Time Complexity: O(1) amortized per operation
 * Space Complexity: O(n)
 * Example: enqueue(1), enqueue(2), dequeue() -> 1, enqueue(3), dequeue() -> 2
 */

import java.util.*

class QueueUsingStacks {
    private val `in` = ArrayDeque<Int>()
    private val out = ArrayDeque<Int>()

    fun enqueue(x: Int) { `in`.push(x) }

    fun dequeue(): Int {
        if (out.isEmpty()) while (`in`.isNotEmpty()) out.push(`in`.pop())
        return out.pop()
    }

    fun peek(): Int {
        if (out.isEmpty()) while (`in`.isNotEmpty()) out.push(`in`.pop())
        return out.peek()
    }

    fun isEmpty() = `in`.isEmpty() && out.isEmpty()
}

fun main() {
    val q = QueueUsingStacks()
    q.enqueue(1); q.enqueue(2)
    println(q.dequeue())
    q.enqueue(3)
    println(q.dequeue())
    println(q.dequeue())
}
